validate/data.go

```diff
@@ -24,36 +33,59 @@ func EntryReader(r dsio.EntryReader) ([]jsonschema.ValError, error) {
 		return nil, fmt.Errorf("error allocating data buffer: %s", err.Error())
 	}
 
+	validated := false
+
 	err = dsio.EachEntry(r, func(i int, ent dsio.Entry, err error) error {
 		if err != nil {
 			return fmt.Errorf("error reading row %d: %s", i, err.Error())
 		}
+
 		err = buf.WriteEntry(ent)
 		if err != nil {
 			return fmt.Errorf("error writing row %d: %s", i, err.Error())
 		}
+		validated = false
+
+		if (i+1)%batchSize == 0 {
+			validated = true
+			if e := buf.Close(); e != nil {
+				return fmt.Errorf("error closing buffer: %s", e.Error())
+			}
+
+			var doc interface{}
+			if err := json.Unmarshal(buf.Bytes(), &doc); err != nil {
+				return fmt.Errorf("error parsing JSON bytes: %s", err.Error())
+			}
+			validationState := jsch.Validate(ctx, doc)
+			valErrors = append(valErrors, *validationState.Errs...)
+
+			buf, err = dsio.NewEntryBuffer(&dataset.Structure{
+				Format: "json",
+				Schema: st.Schema,
+			})
+			if err != nil {
+				return fmt.Errorf("error allocating data buffer: %s", err.Error())
+			}
+		}
+
 		return nil
 	})
 
 	if err != nil {
 		return nil, fmt.Errorf("error reading values: %s", err.Error())
 	}
 
-	if e := buf.Close(); e != nil {
-		return nil, fmt.Errorf("error closing buffer: %s", e.Error())
-	}
-
-	data := buf.Bytes()
-
-	if len(data) == 0 {
-		// TODO (b5): - wut?
-		return nil, fmt.Errorf("err reading data")
-	}
-
-	jsch, err := st.JSONSchema()
-	if err != nil {
-		return nil, err
+	if !validated {
```

Comment:
I'm pretty sure this variable is equivalent to checking if `buf.Byte()` is non-empty. Could you use that check instead and get rid of `validated`?